File: frontend/src/services/usersService.ts
import { api } from './api';
import type { InvestmentProfile, InvestmentProfileUpdate } from '@types/index';

const INVESTMENT_PROFILE_KEY = 'investment_profile';

export const usersService = {
  async updateInvestmentProfile(data: InvestmentProfileUpdate): Promise<InvestmentProfile> {
    const response = await api.patch<InvestmentProfile>('/investment/profile', data);
    localStorage.setItem(INVESTMENT_PROFILE_KEY, JSON.stringify(response.data));
    return response.data;
  },

  getStoredInvestmentProfile(): InvestmentProfile | null {
    const raw = localStorage.getItem(INVESTMENT_PROFILE_KEY);
    if (!raw) return null;
    try {
      return JSON.parse(raw) as InvestmentProfile;
    } catch {
      return null;
    }
  },
};

File: mobile app/lib/services/users_service.dart
import '../core/network/api_client.dart';
import '../core/constants/api_constants.dart';
import '../core/storage/secure_storage.dart';
import '../models/user.dart';

class UsersService {
  final ApiClient _api;
  final SecureStorage _storage;

  UsersService(this._api, this._storage);

  Future<InvestmentProfile> updateInvestmentProfile(InvestmentProfileUpdate data) async {
    final response = await _api.patch(
      ApiConstants.investmentProfile,
      data: data.toJson(),
    );
    final profile = InvestmentProfile.fromJson(response.data as Map<String, dynamic>);
    await _storage.setInvestmentProfile(profile.toJson());
    return profile;
  }

  Future<InvestmentProfile> fetchProfile() async {
    final response = await _api.get(ApiConstants.investmentProfile);
    final profile = InvestmentProfile.fromJson(response.data as Map<String, dynamic>);
    await _storage.setInvestmentProfile(profile.toJson());
    return profile;
  }

  Future<InvestmentProfile?> getStoredProfile() async {
    final data = await _storage.getInvestmentProfile();
    if (data == null) return null;
    return InvestmentProfile.fromJson(data);
  }
}
